### PR TITLE
Improve IndentationHelper

### PR DESCRIPTION
**Pull request by Maxwe11**
- removed string concatenation
- removed char array allocation
- removed delegate and closure allocation

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs [hunk 1]
```diff
@@ -92,19 +91,23 @@ public static SyntaxTrivia GenerateWhitespaceTrivia(IndentationOptions indentati
...
             {
-                indentationString = string.Concat(leadingTrivia[i].ToFullString(), indentationString);
+                if (!trivia.IsKind(SyntaxKind.WhitespaceTrivia))
+                {
+                    break;
+                }
+
+                foreach (char c in trivia.ToFullString())
+                {
+                    indentationCount += c == '\t' ? indentationOptions.TabSize : 1;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: :bulb: The behavior of this is wrong for text like `" \t \t \t \t"` (written as a string literal just so you can see the special characters clearly). Seems like if we're already rewriting the algorithm we could find the issues and fix them too. What do you think?

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs [hunk 2]
```diff
@@ -92,19 +91,23 @@ public static SyntaxTrivia GenerateWhitespaceTrivia(IndentationOptions indentati
...
+                }
             }
 
-            var indentationCount = indentationString.ToCharArray().Sum(c => IndentationAmount(c, indentationOptions));
             return indentationCount / indentationOptions.IndentationSize;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: :question: Shouldn't this be:

```csharp
return (indentationCount + indentationOptions.IndentationSize - 1) / indentationOptions.IndentationSize;
```

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs (hunk 1 above)
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: >The behavior of this is wrong for text like " \t \t \t \t" 

What's wrong with this?

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs (hunk 2 above)
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: `IndentationSize` should be already included as trivia itself in `leadingTrivia` list

Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: Suppose you have a line indented by 3 spaces, but the indentation size is 4. The current implementation will return 0 (i.e. truncation), but my implementation would return 1 (i.e. a ceiling function).

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: That's a bug. Bugs should be fixed separately with appropriate test. Fill in issue :sunglasses:

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs (hunk 1 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: It should return 4, but it currently returns 5. It's counted like you wrote `"    \t\t\t\t"`.

Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/IndentationHelper.cs]: It works as existing code.

**Comment by sharwell** [comment]
I filed issues for the remaining work.

**Pull request merged**